src/v/utils/delta_for.h

```diff
@@ -360,422 +474,67 @@ class deltafor_encoder
     }
 
 private:
-    template<typename T>
-    void _pack_as(const row_t& input) {
-        static_assert(
-          std::is_unsigned<T>::value,
-          "Only unsigned integer can be used as a type parameter");
-        for (unsigned i = 0; i < row_width; i++) {
-            T bits = static_cast<T>(input[i]);
-            uint8_t buf[sizeof(bits)];
-            std::memcpy(&buf, &bits, sizeof(bits));
-            _data.append(buf, sizeof(bits));
+    static_assert(
+      row_width == 16,
+      "pack<N_BITS> assumes row_width = 16, to move N_BITS into a __uint128_t "
+      "and to compute how may bytes of the buffer are non-zero");
+    template<size_t N_BITS>
+    void pack(std::span<const TVal, row_width> input) {
+        if constexpr (N_BITS > 0) {
+            using namespace details::decomp;
+            std::array<uint8_t, serialized_size<N_BITS, row_width>> buff;
+
+            auto end_iter = buff.begin();
+            // step 1: extract and serialize whole bytes, following
+            // decomposition in words
+            constexpr static auto decom = unsigned_decomposition<N_BITS>;
+            [&]<size_t... Is>(std::index_sequence<Is...>) {
+                (
+                  [&] {
+                      for (auto& e : input) {
+                          end_iter = serialize_little_endian<decom[Is].first>(
+                            std::make_unsigned_t<TVal>(e)
+                              >> (8 * decom[Is].second),
+                            end_iter);
+                      }
+                  }(),
+                  ...);
+            }(std::make_index_sequence<decom.size()>{});
+
+            // step 2: pack leftover bits into a single __uint128_t, little
```

Comment:
does the second step corresponds to `_pack*` call? 
`* - [1..7]`

Reply:
yes, it's the first thing i converted. it's a lot of constexpr boilerplate code for 3 lines 517~519 actual work